### PR TITLE
Creating a  single-container  Druid image 

### PR DESCRIPTION
**Pull request by alifouzizadeh**
This custom Docker image for Apache Druid is tailored to our specific needs, addressing limitations we encountered with
the official Druid Docker image which is often optimized for docker-compose or multi-node configurations and it doesn't align with our requirements, e.g. E2E testing. In this PR we will provide a single-container Druid image covering the challenges we face in a multi-node environment. 

I confirm that I have added:

- [x] A comment to every Dockerfile with information about target repository and tag (version).
- [x] Filled in readme on dockerhub and in this repo about how to build the image if some extra steps should be done.

### CODE REVIEW

File: druid/single-container-druid/README.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# Single-container Druid Docker Image
+
+This custom Docker image for Apache Druid is tailored to our specific needs, addressing limitations we encountered with
+the official Druid Docker image. It's particularly useful in environments where the standard Druid Docker setup, which
+is often optimized for docker-compose or multi-node configurations, doesn't align with our requirements, e.g. E2E
+testing.
+
+## Why do we need this image?
+
+- **Custom Configuration**: This image includes additional packages and configurations not present in the official Druid
```
Comment (mpasa) [inline druid/single-container-druid/README.md:10]: `perl` not present for running `start-micro-quickstart`

File: druid/single-container-druid/README.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+```shell
+$ make build DRUID_VERSION=<druid_version> DOCKER_REPO_DOCKERFILE_FOLDER=<child_folder_with_Dockerfile>
+
+```
+
+## run
+
+```shell
+docker run  -p 8081:8081 -p 8082:8082 -p 8083:8083 -p 8888:8888 --name my-druid-container pubnative/single-container-druid:v28.0.0
```
Comment (mpasa) [inline druid/single-container-druid/README.md:32]: Need to close the code block

File: druid/single-container-druid/README.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  complexity of multi-node configurations.
+
+## Build
+
+Use the following command to build the custom Druid Docker image. Replace `<druid_version>` with the desired version of
+Druid, and `<child_folder_with_Dockerfile>` with the path to the folder containing your custom Dockerfile.
+
+```shell
+$ make build DRUID_VERSION=<druid_version> DOCKER_REPO_DOCKERFILE_FOLDER=<child_folder_with_Dockerfile>
+
```
Comment (mpasa) [inline druid/single-container-druid/README.md:26]:
```suggestion
```

**Review by mpasa — CHANGES_REQUESTED** [review]

File: druid/single-container-druid/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,24 @@
+# Image: pubnative/single-container-druid
+
+FROM openjdk:8-jre
+
+RUN apt-get update \
+    && apt-get install -y perl \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
+
+ENV DRUID_VERSION=28.0.0
```
Comment (Konstantinos-Chaitas-PubNative) [inline druid/single-container-druid/Dockerfile:10]: This value here is hardcoded

File: druid/single-container-druid/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,24 @@
+# Image: pubnative/single-container-druid
+
+FROM openjdk:8-jre
```
Comment (Konstantinos-Chaitas-PubNative) [inline druid/single-container-druid/Dockerfile:3]: This value is also hardcoded

File: druid/single-container-druid/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
+
+ENV DRUID_VERSION=28.0.0
+ENV DRUID_DOWNLOAD_URL=https://downloads.apache.org/druid/${DRUID_VERSION}/apache-druid-${DRUID_VERSION}-bin.tar.gz
+
+RUN mkdir /opt/druid
+
+RUN wget -q -O /tmp/druid.tar.gz ${DRUID_DOWNLOAD_URL} \
+    && tar -xzf /tmp/druid.tar.gz -C /opt/druid --strip-components=1 \
```
Comment (Konstantinos-Chaitas-PubNative) [inline druid/single-container-druid/Dockerfile:16]: We typically use the SHA for security reasons

File: druid/single-container-druid/Makefile [hunk 7]
```diff
@@ -0,0 +1,18 @@
+DRUID_VERSION?=28.0.0
```
Comment (Konstantinos-Chaitas-PubNative) [inline druid/single-container-druid/Makefile:1]: Lets update the Makefile following the conventions of AIrflow, Druid, etc..

https://github.com/pubnative/docker-images/pull/200/files#diff-a4396351d4bcc40059cfe8ef85d73ad81c1dd869dfb9a5a3a8570b62073d5499

**Review by Konstantinos-Chaitas-PubNative — CHANGES_REQUESTED** [review]
Can we use the PROD image and pass arguments to it?

**Review by mpasa — APPROVED** [review]

**Comment by alifouzizadeh** [comment]
> Can we use the PROD image and pass arguments to it?

Totally with you on using upstream images. Tried docker-compose first, worked great locally but not on CircleCI. Then moved to the official Druid image, but it wasn’t right for us, it’s more  like for docker-compose or multi-node setups. Modifying the production one wasn’t an option, so I made a custom image instead. Even if we can handle the config and the env, running a single server requires `perl` package and it's not installed in the Official one. I've described this in the Readme file.

**Comment by Konstantinos-Chaitas-PubNative** [comment]
> > Can we use the PROD image and pass arguments to it?
> 
> I already discussed this with @mpasa. Totally with you on using upstream images ( the prod image, in this case). Tried docker-compose first, worked great locally but not on CircleCI. Then moved to the official Druid image, but it wasn’t right for us, it’s more like for docker-compose or multi-node setups. Modifying the production one wasn’t an option, so I made a custom image instead. Even if we can handle the config and the env, running a single server requires `perl` package and it's not installed in the Official one. I've described this in the Readme file.

- What is the problem using docker-compose with CircleCI? I see here they have a guide explaining how it works
https://circleci.com/docs/docker-compose/

- Why do you think that it definitely needs `perl`, I can see this guy for instance doing that without it. My concern is that we will have to update multiple images on version upgrades, which adds maintenance overhead
https://github.com/rilldata/rill/blob/eb98cb2f071668134e8e363d0a0d7ebfc81f7558/scripts/druid-micro.dockerfile#L11

**Review by Konstantinos-Chaitas-PubNative — APPROVED** [review]

**Review by mpasa — APPROVED** [review]

**Pull request merged**